fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/PushDistinctFilterBelowFilterRule.java

```diff
@@ -83,9 +84,10 @@ public PushDistinctFilterBelowFilterRule() {
         super(root);
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public void onMatch(@Nonnull PlannerRuleCall call) {
-        final ExpressionRef<RecordQueryPlan> inner = call.get(innerMatcher);
+        final ExpressionRef<RecordQueryPlan> inner = (ExpressionRef<RecordQueryPlan>)call.get(innerMatcher);
```

Comment:
Please check the new implementation. It uses generics to ensure that if you use a type matcher to filter all the contained expressions you get a reference of the narrowed type rather than the widened type.